File: trees/cppRealizations/diameterofBinaryTree.cpp
#include "btf.h"


class Solution {
public:
    int dfs(TreeNode* curr){
        if(curr == nullptr){
            return 0;
        }
        int left = dfs(curr->left);
        int right = dfs(curr->right);
        result = max(result, left + right);
        return max(left, right) + 1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        // dfs algorithm
        dfs(root);
        return result;
    }
private:
    int result = 0;
};


int main(){
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    cout << "Input:" << endl;
    btf(root);
    cout << "Solution:" << endl;
    cout << Solution().diameterOfBinaryTree(root) << endl;
    return 0;
}

File: trees/diameterofBinaryTree.py
from trees.btf import *


class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.result = 0
        
        # dfs algorithm
        def dfs(curr):
            if not curr:
                return 0
            left = dfs(curr.left)
            right = dfs(curr.right)
            self.result = max(self.result, left + right)
            return 1 + max(left, right)
        
        dfs(root)
        return self.result


if __name__ == '__main__':
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    print("Input:")
    btf(root)
    print("Sulution:")
    print(Solution().diameterOfBinaryTree(root))